Panic when running pull-kubernetes-e2e-kops-aws

<!-- This form is for bug reports and feature requests ONLY! 

If you're looking for help check [Stack Overflow](https://stackoverflow.com/questions/tagged/kubernetes) and the [troubleshooting guide](https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/).
-->

**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug



**What happened**:
See in : https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/55293/pull-kubernetes-e2e-kops-aws/57963/nodelog?junit=junit_08.xml&wrap=on

```
W1108 05:06:09.221535       5 retry_handler.go:54] Inserting delay before AWS request (ec2::DescribeVolumes) to avoid RequestLimitExceeded: 6s
I1108 05:06:09.221638       5 trace.go:76] Trace[1776548750]: "Create /api/v1/persistentvolumes" (started: 2017-11-08 05:06:06.768300566 +0000 UTC m=+680.863858872) (total time: 2.453315887s):
Trace[1776548750]: [2.453315887s] [2.452639812s] END
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1abd33a]

goroutine 915349 [running]:
k8s.io/kubernetes/pkg/cloudprovider/providers/aws.(*CrossRequestRetryDelay).BeforeSign(0xc425fbc300, 0xc42d9b9000)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/retry_handler.go:56 +0x20a
k8s.io/kubernetes/pkg/cloudprovider/providers/aws.(*CrossRequestRetryDelay).BeforeSign-fm(0xc42d9b9000)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws.go:490 +0x34
k8s.io/kubernetes/vendor/github.com/aws/aws-sdk-go/aws/request.(*HandlerList).Run(0xc42d9b9170, 0xc42d9b9000)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/aws/aws-sdk-go/aws/request/handlers.go:195 +0x9d
k8s.io/kubernetes/vendor/github.com/aws/aws-sdk-go/aws/request.(*Request).Sign(0xc42d9b9000, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/aws/aws-sdk-go/aws/request/request.go:343 +0xb0
k8s.io/kubernetes/vendor/github.com/aws/aws-sdk-go/aws/request.(*Request).Send(0xc42d9b9000, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/aws/aws-sdk-go/aws/request/request.go:479 +0x149
k8s.io/kubernetes/vendor/github.com/aws/aws-sdk-go/service/ec2.(*EC2).DescribeVolumes(0xc4245ac758, 0xc42aa80e10, 0x8866b40, 0x4e8581, 0x200334167f1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/aws/aws-sdk-go/service/ec2/api.go:12848 +0x4d
k8s.io/kubernetes/pkg/cloudprovider/providers/aws.(*awsSdkEC2).DescribeVolumes(0xc4245ac760, 0xc42aa80e10, 0x0, 0x42ad38, 0xc42031bc00, 0x7f16b711b000, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws.go:689 +0xc5
k8s.io/kubernetes/pkg/cloudprovider/providers/aws.(*awsDisk).describeVolume(0xc434f59650, 0xc4334167e0, 0x26, 0xc434f59650)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws.go:1515 +0x115
k8s.io/kubernetes/pkg/cloudprovider/providers/aws.(*Cloud).GetVolumeLabels(0xc426679c70, 0xc4334167e0, 0x26, 0x0, 0x0, 0xc425dec188)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/cloudprovider/providers/aws/aws.go:2020 +0x70
k8s.io/kubernetes/plugin/pkg/admission/persistentvolume/label.(*persistentVolumeLabel).findAWSEBSLabels(0xc420b7a6c0, 0xc428371440, 0xc428371440, 0x11, 0x1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/admission/persistentvolume/label/admission.go:136 +0x95
k8s.io/kubernetes/plugin/pkg/admission/persistentvolume/label.(*persistentVolumeLabel).Admit(0xc420b7a6c0, 0x86c8720, 0xc4312236c0, 0x7f16b6b822f8, 0xc420b7a6c0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/admission/persistentvolume/label/admission.go:91 +0x5c4
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission.chainAdmissionHandler.Admit(0xc42013c600, 0xa, 0x10, 0x86c8720, 0xc4312236c0, 0x6, 0x7f16b6b77c01)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission/chain.go:34 +0xf5
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/admission.(*chainAdmissionHandler).Admit(0xc42037e9a0, 0x86c8720, 0xc4312236c0, 0x7f16b6b77c01, 0xc42037e9a0)
	<autogenerated>:1 +0x62
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.createHandler.func1(0x86b9960, 0xc436a15c68, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/create.go:99 +0x15f0
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints.restfulCreateResource.func1(0xc434f58480, 0xc42d52d680)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/installer.go:1080 +0xd5
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.InstrumentRouteFunc.func1(0xc434f58480, 0xc42d52d680)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:168 +0x208
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).dispatch(0xc420715170, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:277 +0xb18
k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).Dispatch(0xc420715170, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:199 +0x57
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x392ac81, 0xe, 0xc420715170, 0xc420479260, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:153 +0x4e0
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.(*director).ServeHTTP(0xc4204d30e0, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	<autogenerated>:1 +0x75
k8s.io/kubernetes/vendor/k8s.io/kube-aggregator/pkg/apiserver.(*proxyHandler).ServeHTTP(0xc42aca1b80, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:91 +0x15f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*pathHandler).ServeHTTP(0xc4347d5d40, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:248 +0x26d
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*PathRecorderMux).ServeHTTP(0xc4222b19d0, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:234 +0xa1
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x392d844, 0xf, 0xc421c4d560, 0xc4222b19d0, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:161 +0x6ad
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.(*director).ServeHTTP(0xc425538180, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	<autogenerated>:1 +0x75
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthorization.func1(0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authorization.go:51 +0x37d
net/http.HandlerFunc.ServeHTTP(0xc42525a5f0, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/usr/local/go/src/net/http/server.go:1918 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithMaxInFlightLimit.func1(0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/maxinflight.go:96 +0x318
net/http.HandlerFunc.ServeHTTP(0xc42138ff80, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/usr/local/go/src/net/http/server.go:1918 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithImpersonation.func1(0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go:49 +0x203a
net/http.HandlerFunc.ServeHTTP(0xc42525a690, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/usr/local/go/src/net/http/server.go:1918 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthentication.func1(0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:79 +0x2b1
net/http.HandlerFunc.ServeHTTP(0xc42525a730, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/usr/local/go/src/net/http/server.go:1918 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request.WithRequestContext.func1(0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/request/requestcontext.go:110 +0xcb
net/http.HandlerFunc.ServeHTTP(0xc4255381c0, 0x7f16b6b71e90, 0xc436a15c58, 0xc430de1500)
	/usr/local/go/src/net/http/server.go:1918 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1(0xc425538280, 0x86bd6a0, 0xc436a15c58, 0xc430de1500, 0xc42f6cfaa0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:93 +0x8d
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:92 +0x1ab
```
/cc @justinsb 

Fixed in #55307

--- a/pkg/cloudprovider/providers/aws/retry_handler.go
+++ b/pkg/cloudprovider/providers/aws/retry_handler.go
@@ -21,6 +21,7 @@ import (
 	"sync"
 	"time"
 
+	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/aws/aws-sdk-go/aws/request"
 	"github.com/golang/glog"
@@ -53,7 +54,15 @@ func (c *CrossRequestRetryDelay) BeforeSign(r *request.Request) {
 	if delay > 0 {
 		glog.Warningf("Inserting delay before AWS request (%s) to avoid RequestLimitExceeded: %s",
 			describeRequest(r), delay.String())
-		r.Config.SleepDelay(delay)
+
+		if sleepFn := r.Config.SleepDelay; sleepFn != nil {
+			// Support SleepDelay for backwards compatibility
+			sleepFn(delay)
+		} else if err := aws.SleepWithContext(r.Context(), delay); err != nil {
+			r.Error = awserr.New(request.CanceledErrorCode, "request context canceled", err)
+			r.Retryable = aws.Bool(false)
+			return
+		}
 
 		// Avoid clock skew problems
 		r.Time = now

